lib/ai/model/tokencount.go

```diff
@@ -0,0 +1,273 @@
+	prompt, err := tc.Prompt.CountAll(ctx)
+	if err != nil {
+		return 0, 0, trace.Wrap(err)
+	}
+	completion, err := tc.Completion.CountAll(ctx)
+	if err != nil {
+		return 0, 0, trace.Wrap(err)
+	}
+	return prompt, completion, nil
+}
+
+// NewTokenCount initializes a new TokenCount struct.
+func NewTokenCount() *TokenCount {
+	return &TokenCount{
+		Prompt:     TokenCounters{},
+		Completion: TokenCounters{},
+	}
+}
+
+// TokenCounter is an interface for all token counters, regardless of the kind
+// of token they count (prompt/completion) or the tokenizer used.
+// TokenCount must be idempotent.
+type TokenCounter interface {
+	TokenCount(ctx context.Context) (int, error)
+}
+
+// TokenCounters is a list of TokenCounter and offers function to iterate over
+// all counters and compute the total.
+type TokenCounters []TokenCounter
+
+// CountAll iterates over a list of TokenCounter and returns the sum of the
+// results of all counters. As the counting process might be blocking/take some
+// time, the caller should set a Deadline on the context.
+func (tc TokenCounters) CountAll(ctx context.Context) (int, error) {
+	var total int
+	for _, counter := range tc {
+		count, err := counter.TokenCount(ctx)
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		total += count
+	}
+	return total, nil
+}
+
+// PromptTokenCounter implements the TokenCounter interface and counts tokens
+// for a given prompt. Tokenization happens on initialization.
+type PromptTokenCounter struct {
+	count int
+}
+
+// TokenCount implements the TokenCounter interface.
+func (tc *PromptTokenCounter) TokenCount(_ context.Context) (int, error) {
+	return tc.count, nil
+}
+
+// NewPromptTokenCounter takes a list of openai.ChatCompletionMessage and
+// computes how many tokens are used by sending those messages to the model.
+func NewPromptTokenCounter(prompt []openai.ChatCompletionMessage) (*PromptTokenCounter, error) {
+	var promptCount int
+	for _, message := range prompt {
+		promptTokens, _, err := defaultTokenizer.Encode(message.Content)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		promptCount = promptCount + perMessage + perRole + len(promptTokens)
+	}
+
+	return &PromptTokenCounter{count: promptCount}, nil
+}
+
+// SynchronousTokenCounter counts completion tokens that have been used by a
+// completion request. This can be used only if the completion request is over
+// and we know the full result. Tokenization happens on initialization.
+type SynchronousTokenCounter struct {
+	count int
+}
+
+// TokenCount implements the TokenCounter interface.
+func (tc *SynchronousTokenCounter) TokenCount(_ context.Context) (int, error) {
+	return tc.count, nil
+}
+
+// NewSynchronousTokenCounter takes the completion request output and
+// computes how many tokens were used by the model to generate this result.
+func NewSynchronousTokenCounter(completion string) (*SynchronousTokenCounter, error) {
```

Comment:
nit: The same as above. Do we need to return the token counter struct where we could only return the count?

Reply:
The point of abstracting the different type of token counting behind an interface was to hide whether token counting was immediate or asynchronous. From the caller pov, everything is a promise of a token count, even if we already computed it.

For prompt tokens we could return the count directly as they are always synchronously countable. For completion tokens that's not the case.

To simplify the prompt and synchronous counters I removed the struct and made them an integer type alias.